src/hotspot/share/memory/metaspace/counters.hpp

```diff
@@ -0,0 +1,183 @@
+ * Copyright (c) 2020 SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_MEMORY_METASPACE_COUNTERS_HPP
+#define SHARE_MEMORY_METASPACE_COUNTERS_HPP
+
+#include "metaprogramming/isSigned.hpp"
+#include "runtime/atomic.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+namespace metaspace {
+
+// We seem to be counting a lot of things which makes it worthwhile to
+// make helper classes for all that boilerplate coding.
+
+// AbstractCounter counts something and asserts overflow and underflow.
+template <class T>
+class AbstractCounter {
+
+  T _c;
+
+  // Only allow unsigned values for now
+  STATIC_ASSERT(IsSigned<T>::value == false);
+
+public:
+
+  AbstractCounter() : _c(0) {}
+
+  T get() const           { return _c; }
+
+  void increment() { increment_by(1); }
+  void decrement() { decrement_by(1); }
+
+  void increment_by(T v) {
+#ifdef ASSERT
+    T old = _c;
+    assert(old + v >= old,
+        "overflow (" UINT64_FORMAT "+" UINT64_FORMAT ")", (uint64_t)old, (uint64_t)v);
+#endif
+    _c += v;
+  }
+
+  void decrement_by(T v) {
+    assert(_c >= v,
+           "underflow (" UINT64_FORMAT "-" UINT64_FORMAT ")",
+           (uint64_t)_c, (uint64_t)v);
+    _c -= v;
+  }
+
+  void reset()                { _c = 0; }
+
+#ifdef ASSERT
+  void check(T expected) const {
+    assert(_c == expected, "Counter mismatch: %d, expected: %d.",
+           (int)_c, (int)expected);
+    }
+#endif
+
+};
+
+// Atomic variant of AbstractCounter.
+template <class T>
+class AbstractAtomicCounter {
+
+  volatile T _c;
+
+  // Only allow unsigned values for now
+  STATIC_ASSERT(IsSigned<T>::value == false);
+
+public:
+
+  AbstractAtomicCounter() : _c(0) {}
+
+  T get() const               { return _c; }
+
+  void increment() {
+    Atomic::inc(&_c);
+  }
+
+  void decrement() {
+#ifdef ASSERT
+    T old = Atomic::load_acquire(&_c);
+    assert(old >= 1,
+        "underflow (" UINT64_FORMAT "-1)", (uint64_t)old);
+#endif
```

Comment:
Okay, I feel stupid now but I do not see it.

I see the non-atomicity, but not how it could cause a false positive assert here. I see that I could miss an assert, but not the other way around.

Reply:
I will remove the asserts. These classes are scheduled to become general utility classes anyway in a followup RFE, so they will be revisited soon.

Reply:
It is not as clear as I though. So no need at all to feel stupid.

> I will remove the asserts. These classes are scheduled to become general utility classes anyway in a followup RFE, so they will be revisited soon.

I'm ok with that.

For the sake of completeness: I'd still think false positives are possible. Please look at the following example:

````
   1	_initialized = 0;
   2	_c = 0;
   3	
   4	Thread A                                  Thread B
   5	
   6	                                          Atomic::inc(&_c)
   7	
   8	                                          <membar StoreLoad|StoreStore>
   9	
  10	                                          _initialized = 1;
  11	
  12	while(_initialized == 0) busy_wait();
  13	
  14	T old = Atomic::load_acquire(&_c);
  15	
  16	assert(old >= 1, "underflow (" UINT64_FORMAT "-1)", (uint64_t)old);
  17	
  18	<fence>
  19	
  20	Atomic::dec(&_c);
````

Loading `old` at line 14 can be done speculatively before the loop, for instance at line 5. There is no barrier that would prevent this. `Atomic::inc` and `Atomic::dec` both have a `<fence>` before and a `<membar StoreLoad|StoreStore>` after the operation. Because of the effect of the `<fence>` in line 18 the `dec()` in line 20 cannot be done before the while in line 12. An underflow is not possible. The assertion can fail though.

So probably you could fix the assertion by adding a fence before the load of the `old` value.